src/Widgets/CalendarChooser.vala

```diff
@@ -0,0 +1,272 @@
+                ((CalendarRow) calrow).selected = false;
+            }
+
+            // Select new row
+            var calrow = row.get_child ();
+            ((CalendarRow) calrow).selected = true;
+        });
+
+        search_entry.activate.connect (() => {
+            foreach (unowned Gtk.Widget child in list_box.get_children ()) {
+                if (child.get_child_visible ()) {
+                    ((Gtk.ListBoxRow) child).activate ();
+                }
+            }
+        });
+
+        search_entry.search_changed.connect (() => {
+            list_box.invalidate_filter ();
+            list_box.unselect_all ();
+            // TODO Should active calendar be re-selected?
+        });
+
+        // Populate list_box
+        foreach (var source in sources) {
+            var calrow = new CalendarRow (source);
+            calrow.margin = 6;
+            calrow.margin_start = 12;
+
+            var row = new Gtk.ListBoxRow ();
+            row.add (calrow);
+            row.show_all ();
+
+            var row_css = """
+                row:selected {
+	                background: mix (#fff, @accent_color, 0.2);
+	                border-radius: 3px;
+                }
+            """;
+            try {
+                var css_provider = new Gtk.CssProvider ();
+                css_provider.load_from_data (row_css);
+                row.get_style_context ().add_provider (css_provider, Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION);
+            } catch (Error e) {
+                warning ("Could not create CSS Provider: %s\nStylesheet:\n%s", e.message, row_css);
+            }
+
+            list_box.add (row);
+
+            if (source.dup_uid () == current_source.dup_uid ()) {
+                list_box.select_row (row);
+                calrow.selected = true;
+            }
+        }
+    }
+
+    [CCode (instance_pos = -1)]
+    private bool filter_function (Gtk.ListBoxRow row) {
+        var search_term = search_entry.text.down ();
+
+        if (search_term in ((CalendarRow)row.get_child ()).label.down ()) {
+            return true;
+        }
+
+        return false;
+    }
+
+    private void header_update_func (Gtk.ListBoxRow row, Gtk.ListBoxRow? before) {
+        var row_location = ((CalendarRow)row.get_child ()).location;
+        if (before != null) {
+            var before_row_location = ((CalendarRow)before.get_child ()).location;
+            if (before_row_location == row_location) {
+                row.set_header (null);
+                return;
+            }
+        }
+
+        var header = new Granite.HeaderLabel (row_location);
+        header.margin = 6;
+        header.margin_bottom = 0;
+
+        row.set_header (header);
+
+        header.show_all ();
+        if (before == null) {
+            header.margin_top = 0;
+        }
+    }
+
+    public void clear_search_entry () {
+        search_entry.text = "";
+    }
+}
+
+public class CalendarRow : Gtk.Grid {
```

Comment:
If this needs to be a public class rather than a private class, it should have its own file

Reply:
Fair enough. Done. One question though: do you think that these CalendarChooser related widgets should go in a subdirectory, or is that just too many and it's fine as is?

Reply:
It's probably okay as is. We can always reorganize if it gets out of hand :)